### ISSUE TITLE
lratsusbst causes bind stack overflow on large lists

### ISSUE THREAD

**Issue by rtoy** [rtoy/maxima#805]
Imported from SourceForge on 2024-07-03 11:32:14
Created by **[l_butler](https://sourceforge.net/u/l_butler/)** on 2021-01-19 20:12:12
Original: https://sourceforge.net/p/maxima/bugs/3706

---

A contrived example:
~~~~
(%i1) load(lrats) $
(%i2) l:makelist(concat('x,i)=concat('x,i+1),i,0,5000) $                
(%i3) lratsubst(l,x0); 
Maxima encountered a Lisp error:
 Condition in MACSYMA-TOP-LEVEL [or a callee]: INTERNAL-SIMPLE-ERROR: Bind stack overflow. 
Automatically continuing. 
To enable the Lisp debugger set *debugger-hook* to nil.   
~~~~
The problem is due to the use of recursion in `ratsubst1`and `fullratsubst1`.

I have attached a patch that re-writes these functions as loops which resolves this problem. The patch includes an iteration limit to prevent infinite loops.

This *resolves* bug [3376](https://sourceforge.net/p/maxima/bugs/3376/), too.
 
Leo

Attachments:
- https://sourceforge.net/p/maxima/bugs/3706/attachment/lrats.mac.diff

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-03 11:32:16
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2021-01-21 17:05:25
Original: https://sourceforge.net/p/maxima/bugs/3706/#18b3

---

Looks good to me, I think you should go ahead. Maybe put one or more test cases into the share test suite.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-03 11:32:20
Created by **[l_butler](https://sourceforge.net/u/l_butler/)** on 2021-01-21 21:40:02
Original: https://sourceforge.net/p/maxima/bugs/3706/#be9e

---

Thanks, Robert.

This is fixed in commit [864fd56e](https://sourceforge.net/p/maxima/code/ci/864fd56e28b7487dafb97c1a7c8dcff3cca78d08/), including tests and documentation updates.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-03 11:32:23
Created by **[l_butler](https://sourceforge.net/u/l_butler/)** on 2021-01-21 21:40:16
Original: https://sourceforge.net/p/maxima/bugs/3706/#3542

---

- **status**: open --> closed
